modules/gdscript/editor/gdscript_docgen.cpp

```diff
@@ -36,40 +36,90 @@ using GDP = GDScriptParser;
 using GDType = GDP::DataType;
 
 static String _get_script_path(const String &p_path) {
-	return vformat(R"("%s")", p_path.get_slice("://", 1));
+	return p_path.trim_prefix("res://").quote();
 }
 
 static String _get_class_name(const GDP::ClassNode &p_class) {
 	const GDP::ClassNode *curr_class = &p_class;
-	if (!curr_class->identifier) { // All inner classes have a identifier, so this is the outer class
+	if (!curr_class->identifier) { // All inner classes have an identifier, so this is the outer class.
 		return _get_script_path(curr_class->fqcn);
 	}
 
 	String full_name = curr_class->identifier->name;
 	while (curr_class->outer) {
 		curr_class = curr_class->outer;
-		if (!curr_class->identifier) { // All inner classes have a identifier, so this is the outer class
+		if (!curr_class->identifier) { // All inner classes have an identifier, so this is the outer class.
 			return vformat("%s.%s", _get_script_path(curr_class->fqcn), full_name);
 		}
 		full_name = vformat("%s.%s", curr_class->identifier->name, full_name);
 	}
 	return full_name;
 }
 
-static PropertyInfo _property_info_from_datatype(const GDType &p_type) {
-	PropertyInfo pi;
-	pi.type = p_type.builtin_type;
-	if (p_type.kind == GDType::CLASS) {
-		pi.class_name = _get_class_name(*p_type.class_type);
-	} else if (p_type.kind == GDType::ENUM && p_type.enum_type != StringName()) {
-		pi.type = Variant::INT; // Only int types are recognized as enums by the EditorHelp
-		pi.usage |= PROPERTY_USAGE_CLASS_IS_ENUM;
-		// Replace :: from enum's use of fully qualified class names with regular .
-		pi.class_name = String(p_type.native_type).replace("::", ".");
-	} else if (p_type.kind == GDType::NATIVE) {
-		pi.class_name = p_type.native_type;
+static void _doctype_from_gdtype(const GDType &p_gdtype, String &r_type, String &r_enum, bool p_is_return = false) {
```

Comment:
`r_type` and `r_enum` means that the arguments must be the `type` (or `return_type`) and `enumeration` (or `return_enum`) properties of `DocData` structures. This is consistent with `EditorHelp::_add_type()` (just GDScript doesn't support `BitField`). `r_enum` cannot contain information about the element type of a typed array.

https://github.com/godotengine/godot/blob/5444afae63f37cb4b890c781e5679a3b7ee5a392/editor/editor_help.h#L161

Reply:
I think I got confused with this line though:

`r_enum += "[]";`

which makes it seem like something is happening with arrays. I guess this is just handling for an array of enums then?

Reply:
Yes, this is a bit of a non-obvious thing. It's because of the following code:

https://github.com/godotengine/godot/blob/5444afae63f37cb4b890c781e5679a3b7ee5a392/editor/editor_help.cpp#L316-L323

Reply:
I think I see. Although won't the `link_t` itself then be `enum_name[]`, which might not work? Do we need to trim the `[]` from the end, perhaps?

I was pretty sleep when I wrote that code, so feel free to change it if you find that it makes more sense some other way! :)

Reply:
This is an interesting question as it affects XML docs.

```
SomeEnum    -> type="int" enum="SomeEnum"
Node        -> type="Node"
Array       -> type="Array"
Array[Node] -> type="Node[]" // Not `type="Array" elem_type="Node"`.
```

I'm not sure which `Array[SomeEnum]` should be according to the logic:

1. `type="SomeEnum[]" enum=""`
2. `type="Array" enum="SomeEnum[]"`
3. `type="int[]" enum="SomeEnum[]"`

But I tested, Option 2 works in Editor Help. Option 3 should also work (the `type` doesn't matter if `enum` is specified). Option 3 is probably the most correct:

```gdscript
enum SomeEnum {}
var a: Array[SomeEnum]

func _run() -> void:
    print(a.get_typed_builtin() == TYPE_INT) # true
    print(a.get_typed_class_name()) # (empty string)
    print(a.get_typed_script()) # <Object#null>
```